### ISSUE TITLE
C++, cpp:var item rejected in in cpp:function

### ISSUE THREAD

**Issue by other-mickk** [sphinx-doc/sphinx#6175]
ping @jakobandersen

I've been using `cpp:var` items inside `cpp:function` items to [help document behaviours etc.](https://github.com/sphinx-doc/sphinx/pull/2538#issuecomment-222520611), and this is now being rejected. For instance:

```rST

.. cpp:function:: template<typename It> void foo(It start, It stop)

    .. cpp:var:: auto const length = std::distance(start, stop)

        The notional distance between the two iterators.

    :complexity: :math:`\Theta(length)`
```

This used to be accepted, with references in `cpp:expr` or `cpp:texpr` linking to the appropriate places and such. This now results in something like:

    WARNING: C++ declarations inside functions are not supported. Parent function is foo()

This looks to be related to the #5191 fix.

**Comment by jakobandersen** [comment]
Good point. Since #5191 the internals for handling overloads is much better so I think it should be possible to relax this again, though with some non-trivial changes.
Note that without the #5191 fix, your example actually works. The fix protects against if you suddenly have a reference to ``length``, say you wrote ``Complexity:  Θ(:cpp:expr:`length`)``. That shouldn't crash now, but I think it would without the fix.
